lib/lf/assembly/assembler.h

```diff
@@ -0,0 +1,219 @@
+      for (int i = 0; i < nrows_loc; i++) {
+        for (int j = 0; j < ncols_loc; j++) {
+          // Add the element at position (i,j) of the local matrix
+          // to the entry at (row_idx[i], col_idx[j]) of the global matrix
+          matrix.AddToEntry(row_idx[i], col_idx[j], elem_mat(i, j));
+        }
+      }  // end assembly local double loop
+    }    // end if(isActive() )
+  }      // end main assembly loop
+}  // end AssembleMatrixLocally
+
+/**
+ * @brief Entity-wise local assembly of a matrix from local matrices
+ *
+ * @return assembled matrix in a format determined by the template argument
+ *         TPMATRIX
+ * @sa  AssembleMatrixLocally(const DofHandler &dof_handler_trial,const
+ * DofHandler &dof_handler_test,ASSEMBLER &assembler, TMPMATRIX &matrix)
+ */
+template <int CODIM, typename TMPMATRIX, class ASSEMBLER>
+TMPMATRIX AssembleMatrixLocally(const DofHandler &dof_handler_trial,
+                                const DofHandler &dof_handler_test,
+                                ASSEMBLER &assembler) {
+  TMPMATRIX matrix{dof_handler_test.GetNoDofs(), dof_handler_trial.GetNoDofs()};
+  AssembleMatrixLocally<CODIM, TMPMATRIX, ASSEMBLER>(
+      dof_handler_trial, dof_handler_test, assembler, matrix);
+  return matrix;
+}
+
+/**
+ * @brief Entity-wise local assembly of a matrix from local matrices
+ *
+ * @return assembled matrix in a format determined by the template argument
+ *         TPMATRIX
+ *
+ * This special version of the function should be used whenever test and trial
+ * space are the same.
+ *
+ * @sa  AssembleMatrixLocally(const DofHandler &dof_handler_trial,const
+ * DofHandler &dof_handler_test,ASSEMBLER &assembler, TMPMATRIX &matrix)
+ */
+
+template <int CODIM, typename TMPMATRIX, class ASSEMBLER>
+TMPMATRIX AssembleMatrixLocally(const DofHandler &dof_handler,
+                                ASSEMBLER &assembler) {
+  return AssembleMatrixLocally<CODIM, TMPMATRIX, ASSEMBLER>(
+      dof_handler, dof_handler, assembler);
+}
+
+/**
+ * @brief entity-local assembly of (right-hand-side) vectors from element
+ * vectors
+ *
+ * @tparam CODIM co-dimension of entities over which assembly should be carried
+ * out
+ * @tparam VECTOR a generic vector type with component access through []
+ * @tparam VECTORASSEMBLER type for objects computing entity-local vectors
+ * @param dof_handler object providing local-to-global dof index mapping, see
+ * DofHandler
+ * @param assembler local assembler object (passed as non-const!)
+ * @param resultvector generic vector for returning the assembled vector
+ *
+ * ### Type requirements for template arguments
+ *
+ * - VECTOR must provide a size() method telling its length and
+ *   read/write access through the [] operator.
+ * - ASSEMBLER must
+ * + offer an `Eval()` method that returns an element vector.
+ * + supply an `isActive()` method for selecting cells to be taken into account
+ * in assembly
+ * + provide a type `ElemVec` suitable for holding an element vector
```

Comment:
This requirement could be relaxed by using auto below.

Reply:
I mean it is not necessary for the Assembler to provide the typedef `ElemVec`

Reply:
I try not to use auto, because obscuring data types is not a good idea.

Reply:
okay I know this is a common critic about `auto`... On the other hand every implementor of an `ASSEMBLER` must now add this type... But I don't mind that much, we can leave it like that...